### PR TITLE
[14.0][BACKPORT] project_forecast_line from 15.0 to 14.0

### PR DESCRIPTION
**Pull request by ntsirintanis**
Removed any relation to `project.project.stage`, as it is not available in 14.0

### CODE REVIEW

**Comment by gurneyalex** [comment]
please add backports of
* #973 
* #974 
* #971 
* #962

**Comment by gurneyalex** [comment]
Regarding project.stage, the equivalement feature on OCA is using project status. Can I suggest that you either replace odoo 15's project stage with OCA's 14.0 project status ? https://github.com/OCA/project/tree/14.0/project_status

**Comment by ntsirintanis** [comment]
> Regarding project.stage, the equivalement feature on OCA is using project status. Can I suggest that you either replace odoo 15's project stage with OCA's 14.0 project status ? https://github.com/OCA/project/tree/14.0/project_status

I'll look into that, thanks for letting me know :+1:

File: project_forecast_line/models/project_task.py
Comment (ntsirintanis) [inline project_forecast_line/models/project_task.py]: @gurneyalex I added this `continue` here, but it is not there in 15.0 version. In my understanding, it should be there. Am I missing something?

**Comment by ntsirintanis** [comment]
@gurneyalex after cherry-picking https://github.com/OCA/project/pull/979 , tests started failing. More specifically, the one test that was introduced by this PR, is failing. Which is strange, because it does not fail on local. Can you please help on this?

**Comment by gjotten** [comment]
> It looks like some users haven't signed our **C**ontributor **L**icense **A**greement, yet. You can read and sign our full Contributor License Agreement here: https://odoo-community.org/page/cla Here is a list of the users:
> 
> * @gjotten (login unknown in OCA database)

Oops, my bad, should be fixed soon since I sent it out just now.

**Comment by ntsirintanis** [comment]
@thomaspaulb @NL66278 I suspect that the reason this one test is failing, is because runboat fails as well in a specific module:

https://runboat.odoo-community.org/api/v1/builds/b2c502af6-71bb-42da-84c8-003b8937bbc7/init-log

Not really sure how to fix this. Locally though the tests are fine.

@gjotten Please go ahead and test this functionally again, if you have time

**Comment by thomaspaulb** [comment]
@ntsirintanis From the looks of the error message in the test run, it's a conflict between `project_task_code` and `project_timesheet_holidays`. As a first line of attack though, could you rebase your branch on the latest `14.0`? You seem to be ~70 commits behind and maybe there is already something in those last commits that solved it, and we don't need to go into it. 14.0 branch itself looks green.

**Comment by thomaspaulb** [comment]
@ntsirintanis Could you rebase once more and we see how the tests run now

File: project_forecast_line/tests/test_forecast_line.py
Comment (thomaspaulb) [inline project_forecast_line/tests/test_forecast_line.py]: @ntsirintanis Currently this test is failing on `odoo.exceptions.ValidationError: Unpaid are only valid starting from 2023-01-01`.

I think it's because of [this line](https://github.com/odoo/odoo/blob/14.0/addons/hr_holidays/data/hr_holidays_data.xml#L49), which is removed in Odoo 15.0.

I think we can patch this test right at the beginning by removing or changing `validity_start` of `hr_holidays.holiday_status_unpaid`?

**Review by thomaspaulb — CHANGES_REQUESTED** [review]

File: project_forecast_line/models/hr_employee.py
Comment (thomaspaulb) [inline project_forecast_line/models/hr_employee.py:33]: Please add backport / cherry-pick of https://github.com/OCA/project/pull/1145

**Comment by gjotten** [comment]
@thomaspaulb remaining problems in CI here are pre-commit as well as big failures in tests. Is this something we can bother you with?

**Review by gjotten — APPROVED** [review]
I can give a functional green light here by the way. What's included in the original 15.0 version works for 14.0 as well.

**Comment by thomaspaulb** [comment]
Hm, the remaining failing test, I think, is an interaction with `project_recalculate`, which doesn't exist in 15.0.

I can't put my finger behind it yet, though

**Comment by gjotten** [comment]
@thomaspaulb you feel like giving this a thumbs-up yourself as well? Or would that be a moral gray area 😄

**Comment by thomaspaulb** [comment]
Well I've finally managed to make the tests green, but:

- [ ] It still needs some Git love, the commit order is now all jambled. We should rebase on latest 15.0, cherry-pick in #1106 and #1145 and then our own commits to backport should come
- [ ] Of our own commits, some, but not all, should be squashed. The loop protection may have to find its way to 15.0 as well, so could be kept separate; and the test-improvement in the other module can be separate as well.
- [ ] Afterwards, it would be really nice if @gurneyalex or @yankinmax could review again as well.

**Comment by thomaspaulb** [comment]
@ntsirintanis Do you feel up to the Git challenge?

**Comment by ntsirintanis** [comment]
@thomaspaulb I squashed something like 10 commits away, will revisit if more squashing is needed.

**Comment by gjotten** [comment]
@OCA/project-service-maintainers 'no stale' should be applied here.

@ntsirintanis I believe you looked into fixing some shortcuts we made here?

**Comment by ntsirintanis** [comment]
@gjotten it's been a while since I pushed anything here, I'll see into making this specific none green soon. And if there are new commits for the module. It looks like that I just need to extend some tests

**Comment by leemannd** [comment]
Hello @ntsirintanis , @gjotten ,
I added the label `no stale`.

I don't know much these modules. From what I understand. You still want to improve unit tests before merging?

**Comment by ntsirintanis** [comment]
> Hello @ntsirintanis , @gjotten , I added the label `no stale`.
> 
> I don't know much these modules. From what I understand. You still want to improve unit tests before merging?

Thanks @leemannd . Yes. The purpose is to make this green and mergeable, ultimately. We have been using it for a while so it is quite stable.

File: project_forecast_line/__manifest__.py
Comment (leemannd) [inline project_forecast_line/__manifest__.py:36]: Open question, as you're using it since  while. Do you think this module is still in `Alpha` ?

Comment (gjotten) [reply project_forecast_line/__manifest__.py:36]: As its sole user internally, I would say yes for the 14.0 version. Also, where we do the backport from in 15.0 also considers this to be in alpha.

File: project_forecast_line/models/hr_employee.py
Comment (yankinmax) [inline project_forecast_line/models/hr_employee.py]: let's get a first (or only one) record from recordset with `fields.first`

**Review by yankinmax — APPROVED** [review]

**Comment by ntsirintanis** [comment]
alright, this is green again

**Comment by ntsirintanis** [comment]
> alright, this is green again

no, I was too hasty to comment

**Comment by ntsirintanis** [comment]
ΟΚ, now truly this one is green

**Comment by ntsirintanis** [comment]
@leemannd thank you very very much for merging this. Can you please reopen this one https://github.com/OCA/project/pull/976 so we can get it merged too?

**Pull request merged**
